### ISSUE TITLE
Multi user mode ?

### ISSUE THREAD

**Issue by behrica** [asreview/asreview#282]
## Feature Request

We have sometimes systematic reviews with thousands of papers to do.

I see a scalability issue with the current ASreview, as it does not have a multi user mode.
Our "traditional" screenings takes weeks to do, with people jumping-in-and-out during that time, until we are finished with all papers.

I can understand, that it is not very easy to think about multi-user in ASReview, due to the active screening approach, in which the next paper to be screened is chosen by the tool.

Eventually I could see multi-user working,
if the model and UI could always show a "ranked list of papers" to a user, on which the user1 could / should "pick" the top one.

If then a second user user2 logs in (while user1 is thinking) , he would as well pick "hist first" one  (which is not the same as the one of user1 ) .

In that the entry point would be a "ranked list of papers"  meant to be worked on by multiple users from top-down.
This list gets then re-shuffled after each submisison,
by the model when it chooses the next one(s) to screen.

Does your active-screening appraoch allow this ?
To produce a ranked list of papers to screen,
or at least something like "the next 10 to screen"

Of course, the software should only permit to pick any article ones... (we work as well with 2 opinions per article ... which would be better, if supported)

I know, this is a big change, but I wanted to start a discussion on this, to get your opinion.

**Comment by behrica** [comment]
I just looked at the API, and it seems that the "ranked list", is not possible.

But maybe something like a "current tasks queue of the project" of a fixed size (10, for example) could be shown,
from which multiple users can "pick" their papers to screen.

If all 10 are done (by multiple users eventually), the tools trains on the 10 and produces a new list of 10.

This seems to be possible from the methods I saw here:
https://asreview.readthedocs.io/en/latest/reference.html

classify, review, query

**Comment by behrica** [comment]
This might help a bit as well to make the app "more asynchronous", which will pay off, the moment you want to have deep learning models, as they would be too slow to assume a "continuous screening" by one user

**Comment by qubixes** [comment]
@behrica Thanks for putting in all the effort.

I agree it would be nice to have a feature like that. I don't think there is a theoretical problem with implementing something like that. The backend could be modified to allow it with some (relatively) small adjustments.

The problem that I see in implementing such a thing right now is that the way that handling the screening is designed in the backend is single user. I think that if you want to implement it smoothly, what you really want is a backend/SQL server that is user aware so to say.

I'm not sure what you mean with "more asynchronous"? The training of the model is currently asynchronous already?

So, great idea, not sure if it is a good idea with the current architecture/design.

**Comment by J535D165** [comment]
I agree with @qubixes that it was designed with a single user perspective. However, we did a multi-user test last week in which both users were training the same model. This was very stable and didn't require a modification of the code.  The downside is that it isn't clear who labeled the item. Two or more opinions per article isn't supported but sounds very interesting. 

I think there are multiple strategies for multi-user screening. Maybe it is good to make an overview of the different strategies and see what is doable to implement.

**Comment by behrica** [comment]
> I'm not sure what you mean with "more asynchronous"? The training of the model is currently asynchronous already?
> 
I mean the user interface is oriented towards a single user, which does the screening one-after-the other in a sequential way. It would get very disrupting, if the training would start to take seconds / minutes each time, after each screening.

In a  "poll oriented" user interface, where user "poll screening tasks from a kind of queue of tasks", this would be less of a problem

**Comment by behrica** [comment]
It is clear that multi-user is a kind of enterprise feature.

It is not clear, if ASReview will go into a direction to be deployed as a single instance on a server and a group of people use the same instance. And it runs then as well 24/7

The current ASReview is still more a single-user application, to be deployed on a desktop computer.
It is more made for a individual scientist, which wants to do a systematic review for his paper.

I have more the perspective of an organisation running several systematic reviews in parallel, by a rather large group of people (hundreds), over a quite long time (each working a few hours per week, when they have time, on the screenings) So each screenings takes up to several weeks.

**Comment by behrica** [comment]
So the screening tools is as well the long term database / archive of all reviews

**Comment by qubixes** [comment]
> I mean the user interface is oriented towards a single user, which does the screening one-after-the other in a sequential way. It would get very disrupting, if the training would start to take seconds / minutes each time, after each screening.
> 
> In a "poll oriented" user interface, where user "poll screening tasks from a kind of queue of tasks", this would be less of a problem

This is already happening in the background though; if the training would take minutes each time, only the first time the user needs to wait (by design), and after that he could continuously screen abstracts, while the computer trains the model in the background. Of course, the model/suggestions isn't updated (almost) immediately as is the case with a quick model.

**Comment by qubixes** [comment]
I understand your perspective, and it would be nice if that kind of deployment would be easy. I have made a docker-compose package last year to do something in that direction. It consisted of a redis server, web server and RQ instances (to run the . I didn't go so far as to create an SQL database for users etc, but it should be very doable. It's more a question of expertise and time in my opinion.

**Comment by wybert** [comment]
It's in 2023, any update for multi user mode?

**Comment by PeterLombaers** [comment]
Were making some progress though not there yet. At the moment were testing a new version of ASReview where multiple users can log in and only see their own projects. This makes it easy to host ASReview on a server and allow multiple people to contribute to the labeling of one dataset (though not at the same moment in time). This will be ready in a month or so would be my guess.

That is not yet the multi-user scenario that was sketched above though, since you can not label with multiple users at the same time. That would be the next step. I'm not entirely sure how long that would take. It might be relatively easy to implement, or we might run into some problems when trying. 

In general I can think of a number of multi-user scenarios:
- Multiple users label the same dataset separately from each other and compare the results at the end (this is already possible of course).
- Multiple users label the same dataset, but not at the same moment in time. (This will be possible with the new update, though keeping track of who is labeled which record requires a bit more work)
- Multilpe users label the same dataset at the same time.

There are probably many other scenarios one can think of. What sort of multi-user scenario are you looking for @wybert ?

**Comment by behrica** [comment]
in our current practice we clearly use "Multilpe users label the same dataset at the same time" without seeing each others labeling decisions.
There is as well a conflict resolution step at the end, and we have as well statistics on this, as a quality criterium on how reliable the SR is overall. This applies for pure manual reviews, of course.

This is as well highly configurable, like how many user  "votes" are required for a "decision to happen":
As I said before, this covers the use case of "long running reviews", where reviewers spent "little time per day / week".

While AsReview seems to focus currently on the "I do a review in a single day (myself)"

**Comment by wybert** [comment]
I want to divide the data into groups and server people can be assigned into one group, for each group people labeling separately simultaneously. Active learning happened after each labeling.
